text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+which [designates a crate as the provider of the final product's panic runtime]
+(less-unwinding).
+Additionally, the standard library's current `panic=unwind` runtime crate,
+`libpanic_unwind`, which is compatible with native C++ style exceptions,
+will be provided under another name (such as `libpanic_native`)
+that guarantees the implementation will remain compatible with C++ exceptions
+(while the implementation of `libpanic_unwind` itself may change
+to no longer maintain that compatibility).
+
+In order for Cargo users to be able
+to specify the C++ compatible `panic_runtime` implementation,
+a new optional value, runtime, will be added to the `profile.foo.panic` option:
+
+```toml
+[profile.dev]
+panic = { 'unwind', runtime = 'native' }
+```
+
+`panic.runtime` may only be specified if the panic strategy is `unwind`.
+And just as there may only be one strategy
+in the dependency graph for a final product,
+there may only be one `runtime`.
+
+For now, the only valid `runtime` keys will be:
+
+* `default` - identical to `panic = 'unwind'` with no `runtime` selected.
+  This will use Rust's default unwinding runtime,
+  unless another crate in the final product's dependency graph
+specifies `panic = 'abort'`,
+  in which case the `abort` strategy will take precedence as usual.
+  which is not guaranteed to be compatible with native exceptions.
+* `native` - as discussed above, this will preserve the behavior of the current
+  implementation of `libpanic_unwind`.
+* `crate` - indicates that a non-`std` crate will use `#![panic_runtime]` to
+  provide the runtime.
+* `self` - indicates that this crate itself provides `#![panic_runtime]`.
+
+If the `native` or `crate` key is specified
+anywhere in a final product's dependency graph,
+no crate in that dependency graph may specify the `panic = abort` strategy;
+this mismatch will cause the build to fail.
+
+The function annotation `#[panic(allowed)]` will only be permitted in crates
+that specify a non-default `panic.runtime`.
+
+Crates that do not use the `profile.foo.panic` option at all
+will remain compatible with any `profile.foo.panic` configuration
+used to generate the final product.
+
+For non-Cargo users, equivalent `rustc` flags will be provided
+(which will be how Cargo itself implements the option).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Unwinding for functions with the `#[unwind(allowed)]` annotation
+is performed as if the function were not marked `extern`.
+This annotation has no effect on functions not marked `extern`.
+It has no observable effect unless the marked function `panic`s
+(e.g. it has no observable effect
+when a function returns normally or enters an infinite loop).
+The LLVM IR for such functions must not be marked `noexcept`.
+
+The compiler will have a new stable flag, `-C panic.runtime`,
+which will be required to enable the `#[unwind(allowed)]` annotation;
```

Comment:
Why does one need to pass a panic-runtime for `#[unwind(allowed)] extern "Rust" fn foo(...) -> ... { ... }` ? AFAICT this function can only be called from other Rust code, and that code must always have a compatible panic-runtime ?

Reply:
Statically-linked Rust code must have a compatible panic-runtime; but AFAIK we can't provide any guarantees about whether dynamically-linked code shares the same runtime (especially with this proposal, which makes it easy to use the new `native` runtime).

Reply:
So how does this work today? IIUC, one can already provide custom panic run-times, and use those to compile libraries and link them, both statically and dynamically.

Reply:
@kornelski The whole problem is that, for dynamically-linkable functions, the compiler _can't know_ whether the caller will have a "compatible" runtime, i.e., it can't know what "compatible runtime" even means. So, from my perspective, the purpose of the `unwind(allowed)` attribute is to let the user say "trust me, the runtime _I've selected_ is compatible with the runtime of any callers of this code." To make this assertion without actually selecting a runtime seems like a logic error from my perspective.

With that said, it does seem like it might be useful to have utility crates that can be used with binaries that use either `unwind.runtime = "native"` or `unwind.runtime = "crate"`. Maybe instead of `crate`, there should be an option that just means "non-default" (compatible with `self` and `native`, but not with `default`).

@gnzlbg Yes, RFC 1513 provides this capability, but it explains that the dependency graph of a binary can only have _one_ `panic` runtime.